torch/csrc/jit/mobile/nnc/context.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <ATen/core/ivalue.h>
+#include <c10/core/ScalarType.h>
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace nnc {
+
+// Specify the requirements on an input tensor.
+// TODO: support input tensor with dynamic shape (PR #54982)
+struct TORCH_API InputSpec {
+  InputSpec() = default;
+
+  // Deserialize the spec from an IValue.
+  explicit InputSpec(const c10::IValue& value);
+
+  // Serialize the spec into an IValue.
+  C10_NODISCARD c10::IValue serialize() const;
+
+  // Check whether the input tensor adheres to the spec.
+  C10_NODISCARD bool validate(const at::Tensor& input) const;
+
+  std::vector<int64_t> sizes_;
+  c10::ScalarType dtype_{c10::ScalarType::Undefined};
+};
+
+// Specify the sizes/dtype/... of output tensor to preallocate the output.
+// TODO: support the case where kernel allocates output tensors dynamically.
+struct TORCH_API OutputSpec {
+  OutputSpec() = default;
+
+  // Deserialize the spec from an IValue.
+  explicit OutputSpec(const c10::IValue& value);
+
+  // Serialize the spec into an IValue.
+  C10_NODISCARD c10::IValue serialize() const;
+
+  // Allocate an output tensor in accordance with the spec.
+  C10_NODISCARD at::Tensor allocate() const;
+
+  std::vector<int64_t> sizes_;
+  c10::ScalarType dtype_{c10::ScalarType::Undefined};
+};
+
+// Hold the temporary buffers / states needed during the execution.
+struct TORCH_API ExecutionState {
+  ExecutionState() = default;
+  ExecutionState(const ExecutionState&) = delete;
+  ExecutionState(ExecutionState&&) = default;
+  ExecutionState& operator=(const ExecutionState&) = delete;
+  ExecutionState& operator=(ExecutionState&&) = default;
+
+  // Preallocated buffers needed by the NNC kernel.
+  std::vector<c10::DataPtr> preallocations_;
+
+  // The NNC kernel expects the following arguments layout:
+  //   input tensor 1
+  //   ...
+  //   input tensor INPUT_NUM
+  //   output tensor 1
+  //   ...
+  //   output tensor OUTPUT_NUM
+  //   parameter tensor 1
+  //   ...
+  //   parameter tensor PARAM_NUM
+  //   temporary buffer 1
+  //   ...
+  //   temporary buffer BUFFER_NUM
+  std::vector<void*> arguments_;
```

Comment:
If arguments are void* can you add a comment to the affect. 
Also for my understanding. What is the temporary buffers? And why there are N of them?

Reply:
Also, I assume this works right now without having to specify size because kernels are statically sized, but this is generic enough for us to get size information at the least? Meaning we can have pair of {tensor_data_ptr, tensor_size_ptr}?

Reply:
I explained temporary buffers in the comment above related the the MemoryPlan.

I agree with these comments - it's just to make the thing work end-to-end with the current AOT compiler prototype so people can iterate on this independently. I think we can make the interface more general in subsequent PRs. We are not even in alpha test so the interface can be changed without much burden.